### PR TITLE
SSO functional testing

### PR DESCRIPTION
**Pull request by patricksimonian**
Stil in draft.

Assertions:
- a user can create a client
- a user can authenticate against client
- a user can get an access token from a client
- a user can refresh an access token from a client

### CODE REVIEW

File: k6/config/index.json [hunk 1]
```diff
@@ -0,0 +1,4 @@
+{
+  "vus": "10",
+  "duration": "30s"
+}
```
Comment (patricksimonian) [inline k6/config/index.json:4]:
```suggestion
}

```

File: k6/index.js [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+
+/**
+ * ENVIRONMENT VARIABLES
+ * SSO_BASE_URL: string
+ * 
+ */
+export default function () {
+  http.get(`${__ENV.SSO_BASE_URL}`);
+  sleep(1);
+}
```
Comment (patricksimonian) [inline k6/index.js:12]:
```suggestion
}

```

File: k6/script.sh [hunk 3]
```diff
@@ -0,0 +1,3 @@
+#!/bin/sh
+
+k6 run index.js --config ./config/index.json
```
Comment (patricksimonian) [inline k6/script.sh:3]:
```suggestion
k6 run index.js --config ./config/index.json

```

**Comment by ShellyXueHan** [comment]
### Updates:
- added token endpoint test case
- added check to verify test case successful
- restructure k6 folder to use shared components
- updated openshift template and added secret for test credentials

### Notes:
@patricksimonian currently job built in tools namespace, and deployed in sandbox testing against sbox.oidc.gov.bc.ca. When token endpoint test pass, you'll see this message from logs: `✓ Token endpoint successfully`

**Comment by ShellyXueHan** [comment]
### Updates -part 2:
- added test cases for SSO API
- updated deployment config to include new env vars
- update script.sh to use env var for k6 runs
- update readme with docker run env var settings

@patricksimonian I'm stuck on some environment issue. The new api test cases are running fine on docker locally, but once running on ocp pod, it'll terminate around the end of first iteration. I've spent 2 hours debugging it already and have no idea what's causing that.... I need your help to take a look what's going on there. Anyway to allow the test to run, I've disabled the api test cases for now.

**Comment by ShellyXueHan** [comment]
@patricksimonian ah nvm figured out! it's resource problem as test cases grow big. I've updated job template with more resource than needed atm as we are still developing them. Need to tweak it afterwards.

anyway! there's log of a total run: https://console.pathfinder.gov.bc.ca:8443/console/project/devops-sso-sandbox/browse/pods/sso-k6-n8zx8?tab=logs

**Review by patricksimonian — COMMENTED** [review]
the logs output the jwt token as a debug log which is probably not ideal!

File: k6/README.md [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+- `SSO_SA_CLIENT_ID`: the Service account client ID
+- `SSO_SA_CLIENT_PASSWORD`: the Service account client password
+
+
+### Configuration
+
+You can modify `config/index.json` to suite your needs. See the [k6 config documentation for options](https://k6.io/docs/using-k6/options)
+
+## Running on OpenShift
+```shell
```
[the remark this answers is not in the document]
Comment (ShellyXueHan) [reply k6/README.md:37]: yes I've added doc for prerequisites. Will need either include the automation in the pre-test or in ansible playbook (preferred).

File: k6/libs/shared.js [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+ * ENVIRONMENT VARIABLES
+ * @param {String} SSO_ENV the environment of the SSO service testing on
+ * @param {String} SSO_REALM the SSO realm testing against
+ * @param {String} SSO_CLIENT the SSO client testing against
+ * @param {String} USER_PASSWORD the test user password 
+ */
+
+// SSO Environment configs:
+export const SSO_CONFIG = {
+  env: (__ENV.SSO_ENV) ? __ENV.SSO_ENV : 'sbox',
```
[the remark this answers is not in the document]
Comment (ShellyXueHan) [reply k6/libs/shared.js]: that's right. Updated!

**Comment by ShellyXueHan** [comment]
### Update:
Originally I thinking that we can use production sso data in sandbox for testing. But having issue now with this approach (see here: https://github.com/bcgov/ocp-sso/issues/108) So for now we cannot simulate DB load.

**Comment by ShellyXueHan** [comment]
### Update:
Have been successfully running those test cases for resource allocation collection. Test cases and oc templates are all updated until we get more test plan. @patricksimonian I can't make you reviewer as you started the PR. Can you take a look and see if it's good to go? Once this PR is in, we can get started with #101

**Comment by ShellyXueHan** [comment]
### Update:
Test cases and doc finished. good to close now.

**Pull request merged**